File: egov/egov-worksweb/src/main/webapp/resources/js/master/scheduleofrate.js
/*
 * eGov suite of products aim to improve the internal efficiency,transparency,
 *    accountability and the service delivery of the government  organizations.
 *
 *     Copyright (C) <2015>  eGovernments Foundation
 *
 *     The updated version of eGov suite of products as by eGovernments Foundation
 *     is available at http://www.egovernments.org
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program. If not, see http://www.gnu.org/licenses/ or
 *     http://www.gnu.org/licenses/gpl.html .
 *
 *     In addition to the terms of the GPL license to be adhered to in using this
 *     program, the following additional terms are to be complied with:
 *
 *         1) All versions of this program, verbatim or modified must carry this
 *            Legal Notice.
 *
 *         2) Any misrepresentation of the origin of the material is prohibited. It
 *            is required that all modified versions of this material be marked in
 *            reasonable ways as different from the original version.
 *
 *         3) This license does not grant any rights to any user of the program
 *            with regards to rights under trademark law for use of the trade names
 *            or trademarks of eGovernments Foundation.
 *
 *   In case of any queries, you can reach eGovernments Foundation at contact@egovernments.org.
 */
$(document).ready(function(){
	if($("#marketRateRow td:eq(2) input.marketratefromdate").val() == "") {
		$("#marketRateRow td input.marketrate").val('0.0');
		$("#marketRateRow").hide();
	}
	else
		$("#marketRateRow").show();
	
	if(parseFloat($(".marketratetable tr td:eq(1)").html()) == 0){
		$(".marketrateview").hide();
	}
	
	$('.textfieldsvalidate').on('input',function(){
		var regexp_textfields = /[^0-9a-zA-Z_@./#&+-/!(){}\",^$%*|=;:<>?`~ ]/g;
		if($(this).val().match(regexp_textfields)){
			$(this).val( $(this).val().replace(regexp_textfields,'') );
		}
	});
	
	if($("#abstractEstimateExists").val() == "true" || $("#workOrderEstimateExists").val() == "true"){
		$($(".sorrate")[0]).attr('readonly', true);
		$($(".sorratefromdate")[0]).attr('readonly', true);
		$($(".deletesorrate")[0]).hide();
	}
});

$('#submitBtn').click(function() {
	if ($('#scheduleOfRateForm').valid()) {
		var isSuccess=true;
		if($("#marketRateRow").is(":not(':hidden')")){
			var marketRateElements = $(".marketrate")
			for (var i=0;i<marketRateElements.length;i++){
				if(!$.isNumeric($(marketRateElements[i]).val()) || parseFloat($(marketRateElements[i]).val()) == 0){
					bootbox.alert($("#marketRateErrorMessage").val());
					isSuccess = false;
					break;
				}
			}
		}
		var sorRateElements = $(".sorrate")
		for (var j=0;j<sorRateElements.length;j++){
			if(!$.isNumeric($(sorRateElements[j]).val()) || parseFloat($(sorRateElements[j]).val()) == 0){
				bootbox.alert($("#sorRateErrorMessage").val());
				isSuccess = false;
				break;
			}
		}
		
		if(!checkSorRateStartDateAndEndDate())
			isSuccess = false;
		
		if(!checkMarketRateStartDateAndEndDate())
			isSuccess = false;
		
		if(!checkSorRateDateBetweenStages())
			isSuccess = false;
		
		if(!checkMarketRateDateBetweenStages())
			isSuccess = false;
		
		if(isSuccess)
			return true;
	}
	return false;
});

$('#btnsearch').click(function() {
	if ($('#scheduleOfRateSearchRequestForm').valid()) {
		callAjaxSearch();
		return true;
	}
	return false;
});

function checkStartDateAndEndDate(tblId,fromDateElements,toDateElements) {
	var checkDateFlag = true;
	$.each(toDateElements, function(index) {
		var fromDateValue = $(fromDateElements[index]).val().split("/");
		var toDateValue = $(this).val().split("/");
		if ($(this).val() != ""
				&& (new Date(fromDateValue[2], fromDateValue[1] - 1,
						fromDateValue[0]) > new Date(toDateValue[2],
						toDateValue[1] - 1, toDateValue[0]))) {
			if(tblId == "tblsorrate")
				bootbox.alert($("#sorInvalidDateRangeErrorMessage").val());
			else
				bootbox.alert($("#marketRateInvalidDateRangeErrorMessage").val());
			checkDateFlag = false;
			return false;
		}
	});
	if (checkDateFlag)
		return true;
	else
		return false;
}

function checkStartDateAndEndDateBetweenStages(tblId){
	var rowcount=$("#"+tblId+" tbody tr").length;
	for(var i=1;i<= rowcount;i++){
		var temp = i-1;
		if(tblId == "tblsorrate"){
			var sorPreviousStageEndDateVal = $("#tempSorRates"+temp+"\\.validity\\.endDate").val();
			var sorCurrentStageStartDateVal = $("#tempSorRates"+i+"\\.validity\\.startDate").val();
			if(sorPreviousStageEndDateVal != "" && sorCurrentStageStartDateVal != undefined){
				var sorPreviousStageEndDate = sorPreviousStageEndDateVal.split("/");
				var sorCurrentStageStartDate = sorCurrentStageStartDateVal.split("/");
				if(new Date(sorPreviousStageEndDate[2], sorPreviousStageEndDate[1] - 1,
						sorPreviousStageEndDate[0]) > new Date(sorCurrentStageStartDate[2],
								sorCurrentStageStartDate[1] - 1, sorCurrentStageStartDate[0])){
						bootbox.alert($("#sorStartEndDateErrorMessage").val());
						return false;
				}
		}
	}else if(tblId == "tblmarketrate"){
			var marketPreviousStageEndDateVal = $("#tempMarketRates"+temp+"\\.validity\\.endDate").val();
			var marketCurrentStageStartDateVal = $("#tempMarketRates"+i+"\\.validity\\.startDate").val();
			if(marketPreviousStageEndDateVal != "" && marketCurrentStageStartDateVal != undefined){
				var marketPreviousStageEndDate = marketPreviousStageEndDateVal.split("/");
				var marketCurrentStageStartDate = marketCurrentStageStartDateVal.split("/");
				if(new Date(marketPreviousStageEndDate[2], marketPreviousStageEndDate[1] - 1,
					marketPreviousStageEndDate[0]) > new Date(marketCurrentStageStartDate[2],
							marketCurrentStageStartDate[1] - 1, marketCurrentStageStartDate[0])){
					bootbox.alert($("#marketRateStartEndDateErrorMessage").val());
					return false;
			}
	}
		}
	}
	return true;
}

function checkSorRateDateBetweenStages(){
	return checkStartDateAndEndDateBetweenStages("tblsorrate");
}

function checkMarketRateDateBetweenStages(){
	if($("#marketRateRow").is(":not(':hidden')")){
		return checkStartDateAndEndDateBetweenStages("tblmarketrate");
	}else
		return true;
}

function checkSorRateStartDateAndEndDate(){
	return checkStartDateAndEndDate("tblsorrate",$(".sorratefromdate"),$(".sorratetodate"));
}

function checkMarketRateStartDateAndEndDate(){
	if($("#marketRateRow").is(":not(':hidden')")){
		return checkStartDateAndEndDate("tblmarketrate",$(".marketratefromdate"),$(".marketratetodate"));
	}else
		return true;
}

var sorRateRowIndex = 0;
function addSorRate() {
	addRow("tblsorrate","sorRateRow");
	if($("#abstractEstimateExists").val() == "true" ||  $("#workOrderEstimateExists").val() == "true"){
		sorRateRowIndex++;
		$($(".sorrate")[sorRateRowIndex]).attr('readonly', false);
		$($(".sorratefromdate")[sorRateRowIndex]).attr('readonly', false);
		$($(".deletesorrate")[sorRateRowIndex]).show();
	}
}

function addMarketRate(){
	if($("#marketRateRow").is(":visible")) 
		addRow("tblmarketrate","marketRateRow");
	$("#marketRateRow").show();
}

function deleteSorRateRow(obj){
	deleteRow(obj,"tblsorrate");
	if($("#abstractEstimateExists").val() == "true" ||  $("#workOrderEstimateExists").val() == "true")
		sorRateRowIndex--;
}

function deleteMarketRateRow(obj){
	var rowcount=$("#tblmarketrate tbody tr").length;
	if(rowcount == 1){
		$("#marketRateRow td").eq(0).find("input").val('');
		$("#marketRateRow td").eq(1).find("input").val('0.0');
		$("#marketRateRow td").eq(2).find("input").val('');
		$("#marketRateRow td").eq(3).find("input").val('');
		$("#marketRateRow").hide();
	}else{
		deleteRow(obj,"tblmarketrate");
	}
}

function addRow(tblId,rowId){
	var rowcount = $("#"+tblId+" tbody tr").length;
	if (rowcount < 30) {
		if ($('#'+rowId+'') != null) {
			// get Next Row Index to Generate
			var nextIdx = 0;
			nextIdx = $("#"+tblId+" tbody tr").length;
			
			// validate status variable for exiting function
			var isValid = 1;// for default have success value 0

			// validate existing rows in table
			$("#"+tblId+" tbody tr").find('input').each(
					function() {
						if (($(this).data('optional') === 0)
								&& (!$(this).val())) {
							$(this).focus();
							bootbox.alert($(this).data('errormsg'));
							isValid = 0;// set validation failure
							return false;
						}
			});

			if (isValid === 0) {
				return false;
			}
			
			// Generate all textboxes Id and name with new index
			$("#"+rowId+"").clone().find("input, errors").each(
					function() {

						if ($(this).data('server')) {
							$(this).removeAttr('data-server');
						}
						
							$(this).attr(
									{
										'id' : function(_, id) {
											return id.replace(/\d+/, nextIdx);
										},
										'name' : function(_, name) {
											return name.replace(/\d+/, nextIdx);
										},
										'data-idx' : function(_,dataIdx)
										{
											return nextIdx;
										}
									});

							// if element is static attribute hold values for
							// next row, otherwise it will be reset
							if (!$(this).data('static')) {
								$(this).val('');
								// set default selection for dropdown
								if ($(this).is("select")) {
									$(this).prop('selectedIndex', 0);
								}
							}
							
							$(this).removeAttr('disabled');
							$(this).prop('checked', false);

					}).end().appendTo("#"+tblId+" tbody");
			
			if(tblId == "tblmarketrate")
				generateSno("marketrate");
			else
				generateSno("sorrate");
			initializeDatePicker();
		}
	} else {
		  bootbox.alert('limit reached!');
	}
}

function deleteRow(obj,tblId) {
    var rIndex = getRow(obj).rowIndex;
    
	var tbl=document.getElementById(''+tblId+'');	
	var rowcount=$("#"+tblId+" tbody tr").length;

    if(tblId != "tblmarketrate" && rowcount<=1) {
		bootbox.alert($("#rowDeleteErrorMessage").val());
		return false;
	} else {
		tbl.deleteRow(rIndex);
		//starting index for table fields
		var idx= 0;
		var sno = 1;
		//regenerate index existing inputs in table row
		jQuery("#"+tblId+" tbody tr").each(function() {
		
				jQuery(this).find("input, select, textarea, errors, span, input:hidden").each(function() {
					var classval = jQuery(this).attr('class');
					if(classval == 'spansno') {
						jQuery(this).text(sno);
						sno++;
					} else {
					jQuery(this).attr({
					      'name': function(_, name) {
					    	  if(!(jQuery(this).attr('name')===undefined))
					    		  return name.replace(/\[.\]/g, '['+ idx +']'); 
					      },
					      'id': function(_, id) {
					    	  if(!(jQuery(this).attr('id')===undefined))
					    		  return id.replace(/\[.\]/g, '['+ idx +']'); 
					      },
					      'class' : function(_, name) {
								if(!(jQuery(this).attr('class')===undefined))
									return name.replace(/\[.\]/g, '['+ idx +']'); 
							},
						  'data-idx' : function(_,dataIdx)
						  {
							  return idx;
						  }
					   });
					}
			    });
				
				idx++;
		});
		if(tblId == "tblmarketrate")
			generateSno("marketrate");
		else
			generateSno("sorrate");
		initializeDatePicker();
		return true;
	}	
}

function callAjaxSearch() {
	drillDowntableContainer = jQuery("#resultTable");
	jQuery('.report-section').removeClass('display-hide');
	reportdatatable = drillDowntableContainer
	.dataTable({
		ajax : {
			url : "/egworks/masters/scheduleofrate-searchdetails",
			type : "POST",
			"data" : getFormData(jQuery('form'))
		},
		"bDestroy" : true,
		'bAutoWidth': false,
		"sDom" : "<'row'<'col-xs-12 hidden col-right'f>r>t<'row'<'col-xs-3'i><'col-xs-3 col-right'l><'col-xs-3 col-right'<'export-data'T>><'col-xs-3 text-right'p>>",
		"aLengthMenu" : [ [ 10, 25, 50, -1 ], [ 10, 25, 50, "All" ] ],
		"oTableTools" : {
			"sSwfPath" : "../../../../../../egi/resources/global/swf/copy_csv_xls_pdf.swf",
			"aButtons" : []
		},
		"fnRowCallback" : function(row, data, index) {
			$('td:eq(0)',row).html(index+1);
			$('td:eq(7)', row).html(
					'<a href="javascript:void(0);" onclick="modifyScheduleOfRate(\''
							+ data.scheduleOfRateId + '\')">Modify</a>');
			return row;
		},
		aaSorting : [],
		columns : [{
				"data" : ""} ,{
				"data" : "sorCode"},{
				"data" : "unitOfMeasure"},{
				"data" : "sorDescription"},{
				"data" : "rate"},{
				"data" : "fromDate"},{
				"data" : "toDate"},{
				"data" : ""
				}]
	});
}

function getFormData($form) {
	var unindexed_array = $form.serializeArray();
	var indexed_array = {};

	$.map(unindexed_array, function(n, i) {
		indexed_array[n['name']] = n['value'];
	});

	return indexed_array;
}

function generateSno(tblName)
{
	var idx=1;
	$("."+tblName+"spansno").each(function(){
		$(this).text(idx);
		idx++;
	});
}

function getRow(obj) {
	if(!obj)
		return null;
	tag = obj.nodeName.toUpperCase();
	while(tag != 'BODY'){
		if (tag == 'TR') 
			return obj;
		obj=obj.parentNode ;
		tag = obj.nodeName.toUpperCase();
	}
	return null;
}

function initializeDatePicker()
{
	$(".datepicker").datepicker({
		format: "dd/mm/yyyy",
		autoclose:true
	});
	try { $(".datepicker").inputmask(); }catch(e){}	
}

function createNewScheduleOfRate(){
	window.location = "scheduleofrate-newform";
}

function modifyScheduleOfRate(scheduleOfRateId) {
	window.location = '/egworks/masters/scheduleofrate-edit/'+scheduleOfRateId;
}

File: egov/egov-works/src/main/java/org/egov/works/masters/entity/ScheduleOfRate.java
/*
 * eGov suite of products aim to improve the internal efficiency,transparency,
 *    accountability and the service delivery of the government  organizations.
 *
 *     Copyright (C) <2015>  eGovernments Foundation
 *
 *     The updated version of eGov suite of products as by eGovernments Foundation
 *     is available at http://www.egovernments.org
 *
 *     This program is free software: you can redistribute it and/or modify
 *     it under the terms of the GNU General Public License as published by
 *     the Free Software Foundation, either version 3 of the License, or
 *     any later version.
 *
 *     This program is distributed in the hope that it will be useful,
 *     but WITHOUT ANY WARRANTY; without even the implied warranty of
 *     MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *     GNU General Public License for more details.
 *
 *     You should have received a copy of the GNU General Public License
 *     along with this program. If not, see http://www.gnu.org/licenses/ or
 *     http://www.gnu.org/licenses/gpl.html .
 *
 *     In addition to the terms of the GPL license to be adhered to in using this
 *     program, the following additional terms are to be complied with:
 *
 *         1) All versions of this program, verbatim or modified must carry this
 *            Legal Notice.
 *
 *         2) Any misrepresentation of the origin of the material is prohibited. It
 *            is required that all modified versions of this material be marked in
 *            reasonable ways as different from the original version.
 *
 *         3) This license does not grant any rights to any user of the program
 *            with regards to rights under trademark law for use of the trade names
 *            or trademarks of eGovernments Foundation.
 *
 *   In case of any queries, you can reach eGovernments Foundation at contact@egovernments.org.
 */
package org.egov.works.masters.entity;

import java.util.ArrayList;
import java.util.Collections;
import java.util.Date;
import java.util.List;

import javax.persistence.CascadeType;
import javax.persistence.Entity;
import javax.persistence.FetchType;
import javax.persistence.GeneratedValue;
import javax.persistence.GenerationType;
import javax.persistence.Id;
import javax.persistence.JoinColumn;
import javax.persistence.ManyToOne;
import javax.persistence.OneToMany;
import javax.persistence.OrderBy;
import javax.persistence.SequenceGenerator;
import javax.persistence.Table;
import javax.persistence.Transient;
import javax.validation.constraints.NotNull;

import org.apache.commons.lang.StringEscapeUtils;
import org.egov.common.entity.UOM;
import org.egov.infra.exception.ApplicationRuntimeException;
import org.egov.infra.persistence.entity.AbstractAuditable;
import org.egov.infra.persistence.entity.component.Period;
import org.egov.infra.persistence.validator.annotation.OptionalPattern;
import org.egov.infra.persistence.validator.annotation.Required;
import org.egov.infra.persistence.validator.annotation.Unique;
import org.egov.infra.utils.StringUtils;
import org.egov.works.utils.WorksConstants;
import org.hibernate.validator.constraints.SafeHtml;
import org.joda.time.LocalDate;

import com.fasterxml.jackson.annotation.JsonIgnore;

@Entity
@Table(name = "EGW_SCHEDULEOFRATE")
@Unique(fields = { "code" }, enableDfltMsg = true)
@SequenceGenerator(name = ScheduleOfRate.SEQ_EGW_SCHEDULEOFRATE, sequenceName = ScheduleOfRate.SEQ_EGW_SCHEDULEOFRATE, allocationSize = 1)
public class ScheduleOfRate extends AbstractAuditable {
    private static final long serialVersionUID = -7797787370112941401L;

    public static final String SEQ_EGW_SCHEDULEOFRATE = "SEQ_EGW_SCHEDULEOFRATE";
    public static final Integer MAX_DESCRIPTION_LENGTH = 100;

    @Id
    @GeneratedValue(generator = SEQ_EGW_SCHEDULEOFRATE, strategy = GenerationType.SEQUENCE)
    private Long id;

    @NotNull
    @SafeHtml
    @OptionalPattern(regex = WorksConstants.ALPHANUMERICWITHALLSPECIALCHARWITHOUTSPACE, message = "sor.code.alphaNumeric")
    private String code;

    @Required(message = "sor.category.not.null")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "SOR_CATEGORY_ID")
    private ScheduleCategory scheduleCategory;

    @NotNull
    @SafeHtml
    @OptionalPattern(regex = WorksConstants.ALPHANUMERICWITHALLSPECIALCHAR, message = "sor.description.alphaNumeric")
    private String description;

    @Required(message = "sor.uom.not.null")
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "UOM_ID")
    private UOM uom;
    
    @JsonIgnore
    @OrderBy("id")
    @OneToMany(mappedBy = "scheduleOfRate", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true, targetEntity = SORRate.class)
    private List<SORRate> sorRates = new ArrayList<SORRate>(0);

    @OrderBy("id")
    @OneToMany(mappedBy = "scheduleOfRate", fetch = FetchType.LAZY, cascade = CascadeType.ALL, orphanRemoval = true, targetEntity = MarketRate.class)
    private List<MarketRate> marketRates = new ArrayList<MarketRate>(0);

    @Transient
    private Double sorRateValue;

    @Transient
    private List<SORRate> tempSorRates = new ArrayList<SORRate>(0);

    @Transient
    private List<MarketRate> tempMarketRates = new ArrayList<MarketRate>(0);

    public String getCode() {
        return code;
    }

    public void setCode(final String code) {
        this.code = code;
    }

    public ScheduleCategory getScheduleCategory() {
        return scheduleCategory;
    }

    public void setScheduleCategory(final ScheduleCategory scheduleCategory) {
        this.scheduleCategory = scheduleCategory;
    }

    public String getDescription() {
        return description;
    }

    public String getDescriptionJS() {
        return StringUtils.escapeJavaScript(description);
    }

    public void setDescription(final String description) {
        this.description = StringEscapeUtils.unescapeHtml(description);
    }

    public UOM getUom() {
        return uom;
    }

    public void setUom(final UOM uomid) {
        uom = uomid;
    }

    public List<SORRate> getSorRates() {
        return sorRates;
    }

    @JsonIgnore
    public List<SORRate> getSorRatesOrderById() {
        Collections.sort(sorRates, (c1, c2) -> c1.getId().compareTo(c2.getId()));
        Collections.reverse(sorRates);
        return sorRates;
    }

    public void setSorRates(final List<SORRate> sorRates) {
        this.sorRates = sorRates;
    }

    public String getSummary() {
        if (description.length() <= MAX_DESCRIPTION_LENGTH)
            return description;
        return first(MAX_DESCRIPTION_LENGTH / 2, description) + "..." + last(MAX_DESCRIPTION_LENGTH / 2, description);

    }

    public String getSummaryJS() {
        return StringUtils.escapeJavaScript(getSummary());
    }

    public String getSorRate() {
        return sorRateValue != null ? sorRateValue.toString() : "";
    }

    public String getScheduleCategorId() {
        return String.valueOf(scheduleCategory.getId());
    }

    protected String first(final int number, final String description) {
        return description.substring(0, number >= description.length() ? description.length() : number);
    }

    protected String last(final int number, final String description) {
        final int begin = description.length() - number;
        return description.substring(begin < 0 ? description.length() : begin, description.length());
    }

    public String getSearchableData() {
        final StringBuilder builder = new StringBuilder();
        builder.append(getCode()).append(" ").append(getDescription());
        return builder.toString();
    }

    public SORRate getRateOn(final Date estimateDate) {
        if (estimateDate == null)
            throw new ApplicationRuntimeException("no.rate.for.date");
        for (final SORRate rate : sorRates)
            if (rate != null && isWithin(rate.getValidity(), estimateDate))
                return rate;
        throw new ApplicationRuntimeException("no.rate.for.date");
    }

    public boolean isWithin(final Period period, final Date dateTime) {
        final LocalDate start = new LocalDate(period.getStartDate());
        LocalDate end = null;
        if (period.getEndDate() != null)
            end = new LocalDate(period.getEndDate());
        final LocalDate date = new LocalDate(dateTime);

        if (end == null)
            return start.compareTo(date) <= 0;
        else
            return start.compareTo(date) <= 0 && end.compareTo(date) >= 0;

    }

    public boolean hasValidRateFor(final Date estimateDate) {
        try {
            final SORRate rate = getRateOn(estimateDate);
            return rate != null;
        } catch (final ApplicationRuntimeException e) {
            return false;
        }
    }

    public void setSorRate(final List<SORRate> sorRates) {
        this.sorRates = sorRates;
    }

    public void addSorRate(final SORRate sorRate) {
        sorRates.add(sorRate);
    }

    public List<MarketRate> getMarketRates() {
        return marketRates;
    }

    public void setMarketRates(final List<MarketRate> marketRates) {
        this.marketRates = marketRates;
    }

    public MarketRate getMarketRateOn(final Date estimateDate) {
        if (estimateDate == null)
            return null;
        for (final MarketRate marketRate : marketRates)
            if (isWithin(marketRate.getValidity(), estimateDate))
                return marketRate;
        return null;
    }

    public boolean hasValidMarketRateFor(final Date estimateDate) {
        final MarketRate marketRate = getMarketRateOn(estimateDate);
        return marketRate != null;
    }

    public void setMarketRate(final List<MarketRate> marketRates) {
        this.marketRates = marketRates;
    }

    public void addMarketRate(final MarketRate marketRate) {
        marketRates.add(marketRate);
    }

    public Double getSorRateValue() {
        return sorRateValue;
    }

    public void setSorRateValue(final Double sorRateValue) {
        this.sorRateValue = sorRateValue;
    }

    @Override
    public Long getId() {
        return id;
    }

    @Override
    protected void setId(final Long id) {
        this.id = id;

    }

    public List<SORRate> getTempSorRates() {
        return tempSorRates;
    }

    public void setTempSorRates(final List<SORRate> tempSorRates) {
        this.tempSorRates = tempSorRates;
    }

    public List<MarketRate> getTempMarketRates() {
        return tempMarketRates;
    }

    public void setTempMarketRates(final List<MarketRate> tempMarketRates) {
        this.tempMarketRates = tempMarketRates;
    }

}